### PR TITLE
HDDS-5014. Upgrade usage primer documentation.

### PR DESCRIPTION
**Pull request by avijayanhwx**
## What changes were proposed in this pull request?
More docs on upgrades.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-5014

## How was this patch tested?
NA

### CODE REVIEW

File: hadoop-hdds/docs/content/design/upgrade-dev-primer.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+### Why does this class exist?
+A typical use case during upgrade is to have multiple versions of a class / method / object and choose them based  on the current layout version at runtime. Before finalizing, an older version is typically needed, and after finalization, a newer version is needed. This class serves this purpose in a generic way. For example, we can create a Factory to create multiple versions of OMRequests sharded by Request Type & Layout Version Supported.
+
+## Upgrade Action (UpgradeActionOm & UpgradeActionHdds)
+Annotation to specify upgrade action run during specific upgrade phases. Each layout feature can optionally define an upgrade action for every supported phase. These are the supported phases of action callbacks.
+
+#### UNFINALIZED_STATE_VALIDATION
+A ‘validation’ action run every time a component is started up with this layout feature being unfinalized.
+
+#### ON_FIRST_UPGRADE_START
```
Comment (swagle) [inline hadoop-hdds/docs/content/design/upgrade-dev-primer.md:62]: Are these actions expected to be idempotent, what is the behavior on failure of this action?

File: hadoop-hdds/docs/content/design/upgrade-dev-primer.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+A backward compatible action run exactly once when an upgraded cluster is detected with this new  layout version.
+
+#### ON_FINALIZE
```
Comment (swagle) [inline hadoop-hdds/docs/content/design/upgrade-dev-primer.md:65]: Same comment as above.

File: hadoop-hdds/docs/content/design/how-to-do-a-nonrolling-upgrade.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+---
+title: How to do a non-rolling upgrade of Ozone
```
Comment (swagle) [inline hadoop-hdds/docs/content/design/how-to-do-a-nonrolling-upgrade.md:2]: These steps should also be a page on the wiki for performing non-rolling upgrades.

File: hadoop-hdds/docs/content/design/upgrade-dev-primer.md (hunk 1 above)
Comment (avijayanhwx) [reply hadoop-hdds/docs/content/design/upgrade-dev-primer.md:62]: Failures are not tolerable in any of these actions. In the onFinalize action failure will cause a finalization failure (can either manually finalize or fix the issue in the action). In the other 2 actions, failure will cause the component to have a failed start.

File: hadoop-hdds/docs/content/design/how-to-do-a-nonrolling-upgrade.md (hunk 3 above)
Comment (avijayanhwx) [reply hadoop-hdds/docs/content/design/how-to-do-a-nonrolling-upgrade.md:2]: Yes, will work with @errose28 to add wiki doc as well.

File: hadoop-hdds/docs/content/design/upgrade-dev-primer.md (hunk 1 above)
Comment (swagle) [reply hadoop-hdds/docs/content/design/upgrade-dev-primer.md:62]: Failures are imminent, as long as user can manually proceed and finalize, can you please add that as a note

Comment (avijayanhwx) [reply hadoop-hdds/docs/content/design/upgrade-dev-primer.md:62]: Thanks @swagle, this will be refactored in a follow up patch.

**Review by swagle — APPROVED** [review]

**Pull request merged**